hydro_cli_maelstrom/src/main.rs

```diff
@@ -0,0 +1,351 @@
+use std::collections::HashMap;
+use std::io::Result;
+use std::net::SocketAddr;
+use std::process::Stdio;
+
+use bytes::Bytes;
+use futures::{SinkExt, StreamExt};
+use serde::{Deserialize, Serialize};
+use serde_json::{json, Value};
+use tokio::io::{stderr, stdin, AsyncBufReadExt, AsyncRead, AsyncWrite, AsyncWriteExt, BufReader};
+use tokio::net::{TcpListener, TcpStream};
+use tokio::process::{ChildStdin, ChildStdout, Command};
+use tokio_util::codec::{FramedRead, FramedWrite, LengthDelimitedCodec};
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub enum ServerPort {
+    TcpPort(SocketAddr),
+}
+
+#[derive(PartialEq, Eq, Serialize, Deserialize)]
+pub enum Direction {
+    Source,
+    Sink,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct PortDefn {
+    pub maelstrom_type: String,
+    pub port_name: String,
+    /// If the port is a sink/source from the perspective of the hydroflow program
+    pub direction: Direction,
+}
+
+/// forwards an in_stream into an out_stream, prefixed by `debug_message`.
+async fn debug_link<R: AsyncRead + Unpin, W: AsyncWrite + Unpin>(
+    in_stream: R,
+    mut out_stream: W,
+    debug_message: String,
+) -> Result<()> {
+    let mut lines = BufReader::new(in_stream).lines();
+    while let Ok(Some(line)) = lines.next_line().await {
+        let text = format!("{debug_message}: {line}\n");
+        out_stream.write_all(text.as_bytes()).await?;
+    }
+    Ok(())
+}
+
+/// Accepts the init payload from maelstrom and responds with an init_ok, returning the node_id of this node
+async fn maelstrom_init() -> Result<String> {
+    let mut init_msg = String::new();
+    BufReader::new(stdin()).read_line(&mut init_msg).await?;
+
+    let v: Value = serde_json::from_str(&init_msg)?;
+    let src = &v["src"];
+
+    let body = &v["body"];
+    let msg_id = &body["msg_id"];
+    let node_id = body["node_id"].as_str().unwrap();
+
+    let response = json!({
+        "src": node_id,
+        "dest": src,
+        "body": {
+            "type": "init_ok",
+            "in_reply_to": msg_id
+        }
+    });
+
+    println!("{}", response);
+
+    Ok(node_id.to_string())
+}
+
+/// Sends setup string which sets up all source ports.
+async fn send_setup_string(
+    source_port_names: &[String],
+    child_stdin: &mut ChildStdin,
+) -> Result<()> {
+    let localhost = HashMap::from([("TcpPort", "127.0.0.1")]);
```

Comment:
To improve the type safety here, we should use the original enum and serialize it with serde.

Reply:
Also, would be great to use Unix sockets when available.